.test-defs/cmd/shoot-maintenance/main.go

```diff
@@ -127,55 +215,17 @@ func init() {
 }
 
 func main() {
-	ctx := context.Background()
-	defer ctx.Done()
-	gardenerConfigPath := fmt.Sprintf("%s/gardener.config", kubeconfigPath)
-
-	shootGardenerTest, err := framework.NewShootGardenerTest(gardenerConfigPath, nil, testLogger)
-	if err != nil {
-		testLogger.Fatalf("Cannot create ShootGardenerTest %s", err.Error())
-	}
-
-	_, shootObject, err := framework.CreateShootTestArtifacts(shootArtifactPath, "", true)
-	if err != nil {
-		testLogger.Fatalf("Cannot create shoot artifact %s", err.Error())
-	}
-
-	shootObject.Name = shootName
-	shootObject.Namespace = projectNamespace
-	shootObject.Spec.Cloud.Profile = cloudprofileName
-	shootObject.Spec.Cloud.Region = region
-	shootObject.Spec.Cloud.SecretBindingRef.Name = secretBindingName
-	shootObject.Spec.Kubernetes.Version = k8sVersion
-	helper.UpdateAnnotations(shootObject)
-	if err := helper.UpdateWorkerZone(shootObject, cloudprovider, zone); err != nil {
-		testLogger.Warnf(err.Error())
-	}
-	if err := helper.UpdateMachineType(shootObject, cloudprovider, machineType); err != nil {
-		testLogger.Warnf(err.Error())
-	}
-	if err := helper.UpdateAutoscalerMin(shootObject, cloudprovider, autoScalerMin); err != nil {
-		testLogger.Warnf(err.Error())
-	}
-	if err := helper.UpdateAutoscalerMax(shootObject, cloudprovider, autoScalerMax); err != nil {
-		testLogger.Warnf(err.Error())
-	}
-	helper.UpdateFloatingPoolName(shootObject, floatingPoolName, cloudprovider)
-	helper.UpdateLoadBalancerProvider(shootObject, loadBalancerProvider, cloudprovider)
-
-	shootGardenerTest.Shoot = shootObject
-
-	shootMaintenanceTest, err := framework.NewShootMaintenanceTest(ctx, shootGardenerTest)
-	if err != nil {
-		testLogger.Fatalf("Failed to create an new Shoot Maintenance test '%v'", err)
-	}
+	defer ctxMain.Done()
+	defer restoreEnvironment()
 
 	found, image, err := v1beta1helper.DetermineMachineImageForName(*shootMaintenanceTest.CloudProfile, shootMaintenanceTest.ShootMachineImage.Name)
 	if err != nil {
-		testLogger.Fatalf("Failed to determine machine image for name: '%s'", err.Error())
+		testLogger.Errorf("Failed to determine machine image for name: '%s'", err.Error())
+		return
```

Comment:
Now the deferred functions are called but the program itself exits with with `0` and therefore does not indicate any failures.

I suggest that we put all the testcode and the deferred cleanup in a separate function that returns an error.
Then we have the cleanup after when an error is returned but still can fail in the main method if an error occurs.

```go
func main() {
  if err := test(); err != nil {
    logger.Fatal(err)
  }
}
func test() error {
  defer cleanup()
  defer cleanup2()

  if err := someTests(); err != nil {
    return err
  }
  
  return nil
}
```

Reply:
Thanks! I did extract it into a separate function. The problem with the above code is, that it would not take errors that happen in the cleanup methods into consideration. Because we are doing non-trivial cleanup we also need to handle that. 
I updated the PR to incorporate you input.